[Serve] Implement Graceful Shutdown

simon-mo opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?
Implement the simplest protocol defined in https://docs.google.com/document/d/1FBkXq8BKfpyxRiDD2BVDLIYZrEBgpiVj4T2m2jj1ikA/edit?usp=sharing

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->
Closes #12297
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

ijrsvt commented on python/ray/serve/tests/test_api.py:
```diff
@@ -871,6 +871,26 @@ def verify_metrics(do_assert=False):
+    # forever. Graceful shutdown is tested in unit test test_backend_worker.py
+    client = serve_instance
+
+    def sleeper(_):
+        while True:
+            time.sleep(1000)
+
+    client.create_backend("sleeper", sleeper)
+    client.create_endpoint("sleeper", backend="sleeper")
+    handle = client.get_handle("sleeper")
+    ref = handle.remote()
+    client.delete_endpoint("sleeper")
+    client.delete_backend("sleeper")
+
+    with pytest.raises(ray.exceptions.RayActorError):
+        ray.get(ref)
```
Hmm, can we adjust the constants for testing? These tests seem quite long

edoakes replied:
+1, make it much shorter for testing

ijrsvt reviewed (commented):
Does this address telling routers at all?

edoakes commented on python/ray/serve/backend_worker.py:
```diff
@@ -410,3 +411,27 @@ async def handle_request(self,
+            if (num_queries_waiting == 0) and (self.num_ongoing_requests == 0):
+                break
+            else:
+                logger.debug(
```
Prefer to make this INFO level

edoakes commented on python/ray/serve/backend_worker.py:
```diff
@@ -410,3 +411,27 @@ async def handle_request(self,
+                break
+            else:
+                logger.debug(
+                    f"Waiting for {DEFAULT_GRACEFUL_SHUTDOWN_LOOP_WAIT_S} "
+                    f"more to shutdown {self.replica_tag} because "
```
```suggestion
                    f"Waiting for an additional {DEFAULT_GRACEFUL_SHUTDOWN_LOOP_WAIT_S}s "
                    f"to shut down replica {self.replica_tag} because "
```

edoakes commented on python/ray/serve/constants.py:
```diff
@@ -49,3 +49,11 @@ def __repr__(self):
     ROUTE_TABLE = auto()
+
+
+#: Backend worker waits for this duration after no work to do then shutdown.
```
```suggestion
#: Duration that backend workers will wait until there is no more work to be done before shutting down.
```

edoakes commented on python/ray/serve/controller.py:
```diff
@@ -308,12 +309,21 @@ async def kill_actor(replica_name_to_use):
-                    # the replica is being restarted and there's no way to tell
-                    # if it successfully killed the worker or not.
-                    ray.kill(replica, no_restart=True)
+                    try:
+                        await asyncio.wait_for(
+                            replica.shutdown.remote(),
+                            timeout=DEFAULT_GRACEFUL_SHUTDOWN_CONTROLLER_WAIT_S
+                        )
+                    except ray.exceptions.RayActorError:
+                        # Actor should have exited.
+                        return
+                    except asyncio.TimeoutError:
+                        # Graceful period passed, kill it forcefully.
+                        logger.debug(
+                            f"{replica_name_to_use} did not shutdown after "
+                            "{DEFAULT_GRACEFUL_SHUTDOWN_CONTROLLER_WAIT_S}, "
+                            "killing.")
+                        ray.kill(replica, no_restart=True)
```
Once we have the asynchronous control loop we can handle this in a better way. Is there any reason not to make this timeout much longer though? In most cases this codepath should never be reached.

edoakes commented on python/ray/serve/tests/test_backend_worker.py:
```diff
@@ -312,6 +315,54 @@ def reconfigure(self, config):
         assert await i == "new_val"
 
 
+async def test_graceful_shutdown(serve_instance, router,
+                                 mock_controller_with_name):
+    class KeepInflight:
+        def __init__(self):
```
We should have a fully end-to-end test using the public API for this as well

simon-mo commented:
@edoakes @ijrsvt I pushed few fixes that addressed the comment. However I found this bug in ray core with async actor https://github.com/ray-project/ray/issues/13049 so I have to change BackendWorker.shutdown to just perform the draining and sleep, while letting the controller to perform the kill.

simon-mo commented on python/ray/serve/config.py:
```diff
@@ -58,6 +60,9 @@ class BackendConfig(BaseModel):
     max_concurrent_queries: Optional[int] = None
     user_config: Any = None
 
+    experimental_graceful_shutdown_wait_loop_s: PositiveFloat = 2.0
+    experimental_graceful_shutdown_timeout_s: PositiveFloat = 20.0
+
```
Made shutdown parameters per backend instead of the global constants. (in k8s, `terminationGracePeriodSeconds` is per pod config anyway)

simon-mo commented on python/ray/serve/controller.py:
```diff
@@ -361,22 +377,27 @@ async def _check_currently_stopping_replicas(self) -> bool:
-                await self._check_currently_starting_replicas()
-                or await self._check_currently_stopping_replicas())
+                delta = time.time() - start
+                logger.warning(
```
fixed the warning message to make more sense.

ijrsvt commented on python/ray/serve/router.py:
```diff
@@ -226,7 +250,14 @@ async def assign_request(
+            endpoint_status = await self._pending_endpoints[endpoint]
+            if endpoint_status == _PendingEndpointFound.REMOVED:
+                raise RayServeException(
+                    f"Endpoint {endpoint} was removed therefore this request "
+                    "cannot be completed.")
```
```suggestion
                    f"Endpoint {endpoint} was removed. This request "
                    "cannot be completed.")
```
nit

ijrsvt commented on python/ray/serve/config.py:
```diff
@@ -30,25 +30,27 @@ class BackendMetadata:
+        user_config (Optional[Any]): Arguments to pass to the reconfigure
+            method of the backend. The reconfigure method is called if
+            user_config is not None.
+        experimental_graceful_shutdown_wait_loop_s (Optional[float]): Duration
+            that backend workers will wait until there is no more work to be
+            done before shutting down. Defaults to 2s.
```
Maybe mention that  `>= time_for_router_long_polling_update`

simon-mo replied:
hmm this value isn't expected to be updated, it should be set by create_backend and populated on replica construction.

ijrsvt reviewed (commented):
Looks good, one quick question: Will this dramatically increase the testing time for Serve because most tests have the following fixture: 
```
@pytest.fixture
def serve_instance(_shared_serve_instance):
...
delete_endpoints
delete_backends
```

simon-mo commented:
@ijrsvt, added a `delete_backend(..., force=False)` argument to add override and make tests faster.

ijrsvt reviewed (approved):
Looks good!

architkulkarni commented:
Still looking this over, but it looks like `test_backend_worker.py::test_graceful_shutdown' is failing right now

simon-mo commented:
@architkulkarni pushed a fixed! thx

architkulkarni reviewed (approved):
Looks great!
